### PR TITLE
Improve error messages during validation.

### PR DESCRIPTION
**Pull request by davidbtucker**

### CODE REVIEW

File: confgenerator/config.go [hunk 1]
```diff
@@ -82,7 +82,37 @@ func (ve validationError) StructField() string {
 	return parts[0]
 }
 
-// TODO: Implement validationError.Error() for better error messages.
+func (ve validationError) Error() string {
+	e := ve.FieldError
```
[the remark this answers is not in the document]
Comment (davidbtucker) [reply confgenerator/config.go]: Done.

File: confgenerator/config.go [hunk 2]
```diff
@@ -82,7 +82,37 @@ func (ve validationError) StructField() string {
...
+
+	switch e.Tag() {
+	case "duration":
+		return fmt.Sprintf("%s must be a duration of at least %s", e.Field(), e.Param())
```
[the remark this answers is not in the document]
Comment (davidbtucker) [reply confgenerator/config.go]: Done.

File: confgenerator/config.go [hunk 3]
```diff
@@ -82,7 +82,37 @@ func (ve validationError) StructField() string {
...
+		return fmt.Sprintf("%s must end with %q", e.Field(), e.Param())
+	case "ip":
+		return fmt.Sprintf("%s must be an IP address", e.Field())
+	case "oneof":
+		return fmt.Sprintf("%s must be one of [%s]", e.Field(), e.Param())
+	case "required":
+		return fmt.Sprintf("%s is a required field", e.Field())
+	case "startsnotwith":
+		switch {
+		case strings.HasPrefix(e.StructField(), "Pipelines"):
```
[the remark this answers is not in the document]
Comment (davidbtucker) [reply confgenerator/config.go]: These are the values:

    pipelines[lib:pipeline_1]
    Pipelines[lib:pipeline_1]
    processors[lib:processor_1]
    receivers[lib:receiver_1]

We could use the bit before `s[...]` if you think that would be clearer.

File: confgenerator/testdata/invalid/linux/logging-pipeline_reserved_id_prefix/golden_error [hunk 4]
```diff
@@ -1,4 +1,4 @@
-the agent config file is not valid. detailed error: [10:21] Key: 'LoggingService.Pipelines[lib:pipeline_1]' Error:Field validation for 'Pipelines[lib:pipeline_1]' failed on the 'startsnotwith' tag
+the agent config file is not valid. detailed error: [10:21] pipeline name must not start with "lib:"
```
[the remark this answers is not in the document]
Comment (davidbtucker) [reply confgenerator/testdata/invalid/linux/logging-pipeline_reserved_id_prefix/golden_error:1]: Do mean the `The agent...` or `Pipeline name...` or both?

**Review by davidbtucker — COMMENTED** [review]
Thanks, need clarification on two of the comments.

File: confgenerator/config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (quentinmit) [reply confgenerator/config.go]: Can you just say "%s must not start with %q", e.Field(), e.Param() ? I think `processors[lib:processor_1] must not start with "lib:"` is a perfectly fine message and keeps the code simple.

File: confgenerator/testdata/invalid/linux/logging-pipeline_reserved_id_prefix/golden_error (hunk 4 above)
[the remark this answers is not in the document]
Comment (quentinmit) [reply confgenerator/testdata/invalid/linux/logging-pipeline_reserved_id_prefix/golden_error:1]: I mean `the agent config...`

This is coming from `UnmarshalYamlToUnifiedConfig` in `config.go`. That function should just return the bare err, and then you can add a well-formatted prefix all the way out at the `log.Fatal` call in `cmd/google_cloud_ops_agent_engine/main.go`.

File: confgenerator/config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidbtucker) [reply confgenerator/config.go]: Done.

File: confgenerator/testdata/invalid/linux/logging-pipeline_reserved_id_prefix/golden_error (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidbtucker) [reply confgenerator/testdata/invalid/linux/logging-pipeline_reserved_id_prefix/golden_error]: Done.

**Review by quentinmit — APPROVED** [review]
Looks like a significant improvement to me, thanks!

**Review by qingling128 — APPROVED** [review]
LGTM

**Pull request merged**